### PR TITLE
Like button

### LINKED ISSUE #125: Like Button
- [x] Implement a like button component with a counter, which takes a entityType and entityId as input
- [x] When render, it load the existing count from the trpc API
- [x] Upon Click, counter is incremented,
- [x] Ensure only one user can like once

### PR DESCRIPTION
**Pull request by LovedeepSingh19**
You can test the like button in marketplace packages.

Screenshot 2023-12-28 at 12 28 10 PM
--
Screenshot 2023-12-28 at 12 29 32 PM

Screenshot 2024-01-09 at 12 18 30 PM

### CODE REVIEW

File: apps/factory/prisma/schema.prisma [hunk 1]
```diff
@@ -222,3 +222,28 @@ model PromptLog {
   createdAt         DateTime          @default(now()) @map("created_at")
   updatedAt         DateTime          @updatedAt @map("updated_at")
 }
+
+model Like {
+  id         String     @id @default(uuid())
+  likes      LikeUser[]
+  likesCount Int
```
Comment (devilankur18) [inline apps/factory/prisma/schema.prisma:229]: add default value to 0

File: apps/factory/src/components/marketplace/like_button.tsx [hunk 2]
```diff
@@ -1,14 +1,76 @@
...
+  const [loading, setLoading] = useState(false);
+  const [LikeId, setLikeId] = useState<string>("");
+
+  useEffect(() => {
+    data?.likes.some((like) => {
+      if (like.userId === sessionData?.user.id) {
+        setHasLiked(true);
+        setLikeId(like.likeId!);
+      }
+    });
```
Comment (devilankur18) [inline apps/factory/src/components/marketplace/like_button.tsx:33]: @LovedeepSingh19 this is not scalable, you can't pull all the likes to frontend. The user id need to send to the backend and check in db. This need to be changed.

File: apps/factory/src/components/marketplace/like_button.tsx [hunk 3]
```diff
@@ -1,14 +1,76 @@
...
+        },
+      },
+    );
+
+    if (hasLiked) {
+      setCounter(counter! - 1);
+      setHasLiked(!hasLiked);
+    } else {
+      setCounter(counter! + 1);
+      setHasLiked(!hasLiked);
```
Comment (devilankur18) [inline apps/factory/src/components/marketplace/like_button.tsx:71]: We need to discuss this. The hasLiked should not be sent back to the like. Like action should be done and completlely handled with out hasLiked information from frontend.

File: apps/factory/src/server/api/routers/likes.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        .$transaction(async (prisma) => {
+          if (hasLiked) {
+            // Delete the existing like from LikeUser table
+            await prisma.likeUser.deleteMany({
+              where: {
+                userId: ctx.session?.user.id,
+                likeId: LikeId,
+              },
+            });
+
```
Comment (devilankur18) [inline apps/factory/src/server/api/routers/likes.ts:20]: Create seperate action for unlike and like, don't use the same mutation. We need to enable caching and other mechanism for the same in the near future.

**Review by devilankur18 — CHANGES_REQUESTED** [review]

File: apps/factory/src/components/marketplace/like_button.tsx (hunk 2 above)
Comment (LovedeepSingh19) [reply apps/factory/src/components/marketplace/like_button.tsx:33]: You told me to put that code in frontend. Previously in meet I showed you the code and 1st line was checking that same thing

**Review by devilankur18 — APPROVED** [review]

**Pull request closed without merging**